transport-native-epoll/src/main/java/io/netty/channel/epoll/TunTapPacketLogger.java

```diff
@@ -0,0 +1,217 @@
+                TunTapPacket.getProtocolName(packet.protocol()),
+                packetData.readableBytes());
+
+        // If the packet is an Ethernet frame...
+        if (isEtherFrame) {
+
+            // Generally a sender doesn't specify a protocol when sending an Ethernet frame, since
+            // Linux doesn't require it.  Therefore extract the protocol from the Ethernet frame data.
+            protocol = packetData.getUnsignedShort(packetData.readerIndex() + 12);
+
+            // Print the Ethernet header
+            formatEthernetHeader(packetData, 0, protocol, formatter);
+
+            // Adjust where we look for the IP header within the packet.
+            ipPacketOffset = 14;
+        }
+
+        // If the packet contains an IPv4 or IPv6 packet, print the appropriate header information.
+        if (protocol == TunTapPacket.PROTOCOL_IPV6) {
+            formatIPv6Header(packetData, ipPacketOffset, formatter);
+        } else if (protocol == TunTapPacket.PROTOCOL_IPV4) {
+            formatIPv4Header(packetData, ipPacketOffset, formatter);
+        }
+
+        // Print the full contents of the packet.
+        if (logPacketData) {
+            formatter.format("%s", ByteBufUtil.prettyHexDump(packetData));
+        }
+
+        formatter.close();
+
+        return out.toString();
+    }
+
+    private static void formatEthernetHeader(ByteBuf packetData, int headerOffset, int protocol, Formatter formatter) {
+        int readerIndex = packetData.readerIndex();
+        byte[] srcMACAddr = new byte[6];
+        byte[] destMACAddr = new byte[6];
+
+        packetData.getBytes(readerIndex + headerOffset + 6, srcMACAddr);
+        packetData.getBytes(readerIndex + headerOffset + 0, destMACAddr);
+
+        formatter.format("  Ethernet: src ");
+        formatMAC48(srcMACAddr, formatter);
+        formatter.format(", dest ");
+        formatMAC48(destMACAddr, formatter);
+        formatter.format(", protocol %s\n", TunTapPacket.getProtocolName(protocol));
+    }
+
+    private static void formatIPv4Header(ByteBuf packetData, int headerOffset, Formatter formatter) {
+        int readerIndex = packetData.readerIndex();
+        String srcAddr, destAddr;
+
+        try {
+            byte[] srcAddrData = new byte[4];
+            packetData.getBytes(readerIndex + headerOffset + 12, srcAddrData);
+            srcAddr = InetAddress.getByAddress(srcAddrData).getHostAddress();
+        } catch (Exception ex) {
+            srcAddr = "(invalid)";
+        }
+
+        try {
+            byte[] destAddrData = new byte[4];
+            packetData.getBytes(readerIndex + headerOffset + 16, destAddrData);
+            destAddr = InetAddress.getByAddress(destAddrData).getHostAddress();
+        } catch (Exception ex) {
+            destAddr = "(invalid)";
+        }
+
+        int transportProtocol = packetData.getUnsignedByte(readerIndex + headerOffset + 9);
+
+        formatter.format("  IPv4: src %s, dest %s, protocol %s\n",
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) %n should be used in place of \n to produce the platform-specific line separator. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2275)